sdk/trace/ratelimit/ratelimit.go

```diff
@@ -0,0 +1,218 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ratelimit // import "go.opentelemetry.io/otel/sdk/trace/ratelimit"
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"go.opentelemetry.io/otel/sdk/trace"
+)
+
+type config struct {
+	interval time.Duration
+	nowfunc  func() time.Time
+}
+
+type Option interface {
+	apply(*config)
+}
+
+type window struct {
+	// start is the beginning of this window
+	start time.Time
+
+	// low is the lower power-of-two probability
+	low trace.Sampler
+	// high is the higher power-of-two probability
+	high trace.Sampler
+
+	// lowProb is the probability of sampling at the lowProb
+	lowProb float64
+
+	count   int64
+	compute sync.Once
+}
+
+type Sampler struct {
+	targetCount int
+	interval    time.Duration
+	nowfunc     func() time.Time
+
+	current atomic.Value
+
+	priorCount    int64
+	priorDuration time.Duration
+}
+
+const (
+	DefaultProbability = 1
+	DefaultInterval    = 10 * time.Second
+	MinInterval        = 10 * time.Millisecond
+	MinRate            = 0.00001
```

Comment:
Curious if it's possible to compute MinRate from MaxRate instead of a constant. And similarly, are there any guarantees or trends on the actual rate vs the MaxRate (always within x% of the MaxRate, never goes over, etc).

Reply:
Discussed in the main review thread.